### PR TITLE
[WAVE 7] Hold Requests: Individual Money Request Page, Reason Interstitial, Banner

### PR DESCRIPTION
**Pull request by BartoszGrajdek**
### Fixed Issues

$ https://github.com/Expensify/App/issues/31300

### Tests

1. Send a money request
2. Open this newly created money request page
3. In top right corner click three dots and "Hold request" button (in a workspace people able to put the request on hold are: admins, approvers that the request has been sent to, and owner of the money request, in personal IOU all members should be able to put the request on hold)
4. Check if the hold reason popup was opened 
5. Make sure that an error is being displayed when submitting a hold reason without input being filled
6. Fill in the hold reason input and submit
7. Check if the Hold Banner on top has appeared and a system message with a comment has been sent
8. Log into another account that is a member of the money request and check if everything is displayed properly
9. On personal IOU requests only the person who has put the request on hold be able to see/use the "Unhold request" button in three dots menu
10. On workspace money request admin, the approver that the request has been sent to and the person that put the money request on hold should be able to see/use the "Unhold request" button 
11. Click the "Unhold request" button and check if a system message has been sent saying that the request has been unheld
12. Check different accounts with different permissions, and make sure that the "Hold request" & "Unhold request" buttons are being displayed and working for proper accounts (according to the description above)

### Offline tests

1. Send a money request (this and the second step can be both done in offline and online mode to check permissions)
2. Open this newly created money request page
4. In the top right corner click three dots and the "Hold request" button (in a workspace people able to put the request on hold are: admins, approvers that the request has been sent to, and owner of the money request, in personal IOU all members should be able to put the request on hold)
5. Check if the hold reason popup was opened 
6. Make sure that an error is being displayed when submitting a hold reason without input being filled
7. Fill in the hold reason input and submit
8. Check if the Hold Banner on top has appeared and a system message with a comment has been sent
10. On personal IOU requests only the person who has put the request on hold be able to see/use the "Unhold request" button in three dots menu
11. On workspace money request admin, the approver that the request has been sent to and the person that put the money request on hold should be able to see/use the "Unhold request" button 
12. Click the "Unhold request" button and check if a system message has been sent saying that the request has been unheld
13. Check different accounts with different permissions, and make sure that the "Hold request" & "Unhold request" buttons are being displayed and working for proper accounts (according to the description above)

### QA Steps

1. Send a money request
2. Open this newly created money request page
3. In top right corner click three dots and "Hold request" button (in a workspace people able to put the request on hold are: admins, approvers that the request has been sent to, and owner of the money request, in personal IOU all members should be able to put the request on hold)
4. Check if the hold reason popup was opened 
5. Make sure that an error is being displayed when submitting a hold reason without input being filled
6. Fill in the hold reason input and submit
7. Check if the Hold Banner on top has appeared and a system message with a comment has been sent
8. Log into another account that is a member of the money request and check if everything is displayed properly
9. On personal IOU requests only the person who has put the request on hold be able to see/use the "Unhold request" button in three dots menu
10. On workspace money request admin, the approver that the request has been sent to and the person that put the money request on hold should be able to see/use the "Unhold request" button 
11. Click the "Unhold request" button and check if a system message has been sent saying that the request has been unheld
12. Check different accounts with different permissions, and make sure that the "Hold request" & "Unhold request" buttons are being displayed and working for proper accounts (according to the description above)

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/41169160/35354e95-7ea9-4e09-866a-824c139adba9

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/41169160/f1d0ef01-6b5d-4a43-a55e-1a752ac4afb7

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/41169160/ff766256-f8cc-4dfb-bc3d-587edd7f25e4

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/41169160/99694fc0-9a1c-4703-9532-bf0fe135144f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/41169160/e882419f-6c3b-4f26-ad86-ad8029bed72e

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/41169160/30820339-5add-4fa3-b6cb-6bbea0457c95

</details>

### CODE REVIEW

File: src/languages/es.ts [hunk 1]
```diff
@@ -654,14 +654,26 @@ export default {
...
+        approveOnly: 'Solo aprobar',
+        hold: 'Bloqueada',
+        holdEducationalTitle: 'Esta solicitud está',
+        whatIsHoldTitle: '¿Qué es Bloquear?',
+        whatIsHoldExplain: 'Bloquear es nuestra forma de agilizar la colaboración financiera. ¡"Rechazar" es tan duro!',
+        holdIsTemporaryTitle: 'Bloquear suele ser temporal',
+        holdIsTemporaryExplain: 'Se utiliza bloquear para aclarar confusión o aclarar un detalle importante antes del pago, no es permanente.',
         deleteHoldTitle: 'Eliminar lo que no se pagará',
-        deleteHoldExplain: 'En el raro caso de que algo se ponga en hold y no se pague, la persona que solicita el pago debe eliminarlo.',
+        deleteHoldExplain: 'En el raro caso de que algo se bloquear y no se pague, la persona que solicita el pago debe eliminarlo.',
```
Comment (kowczarz) [inline src/languages/es.ts:676]: Just not to skip this, have you confirmed translations on #expensify-open-source?

**Review by kowczarz — APPROVED** [review]

File: src/languages/es.ts (hunk 1 above)
Comment (BartoszGrajdek) [reply src/languages/es.ts:676]: This actually comes from Expensify internal engineers 😄

**Review by robertjchen — APPROVED** [review]

**Comment by robertjchen** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

File: src/pages/iou/HoldReasonPage.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            >
+                <Text style={[styles.textHeadline, styles.mb6]}>{translate('iou.explainHold')}</Text>
+                <View>
+                    <InputWrapper
+                        InputComponent={TextInput}
+                        inputID={INPUT_IDS.COMMENT}
+                        valueType="string"
+                        name="comment"
+                        defaultValue={undefined}
+                        label="Reason"
```
Comment (MonilBhavsar) [inline src/pages/iou/HoldReasonPage.tsx:87]: @BartoszGrajdek We missed translate here and that created https://github.com/Expensify/App/issues/36826
Do you please mind pushing a quick fix?

**Comment by MonilBhavsar** [comment]
We have this another issue where composer stays focussed when Hold Request RHP is open. If anyone is interested in taking it 🙏 https://github.com/Expensify/App/issues/36829

File: src/libs/ReportUtils.ts [hunk 3]
```diff
@@ -3492,6 +3496,72 @@ function buildOptimisticRenamedRoomReportAction(newName: string, oldName: string
...
+    return {
+        reportActionID: NumberUtils.rand64(),
+        actionName: CONST.REPORT.ACTIONS.TYPE.HOLD,
+        pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD,
+        actorAccountID: currentUserAccountID,
+        message: [
+            {
+                type: CONST.REPORT.MESSAGE.TYPE.TEXT,
+                style: 'normal',
+                text: `held this money request with the comment: ${comment}`,
```
Comment (MonilBhavsar) [inline src/libs/ReportUtils.ts:3513]: And we're also missing translate here and at L3548 below that caused https://github.com/Expensify/App/issues/36836.

**Comment by dragnoir** [comment]
> We have this another issue where composer stays focussed when Hold Request RHP is open. If anyone is interested in taking it 🙏 #36829

I created a proposal https://github.com/Expensify/App/issues/36829#issuecomment-1953775087

**Comment by BartoszGrajdek** [comment]
@MonilBhavsar I can create a PR fixing all the regressions you've mentioned, just let me know if you treat [this](https://github.com/Expensify/App/issues/36829) as a regression as well - I think it is one, but I can already see some proposals there, so I don't want to cause any confusion

**Comment by BartoszGrajdek** [comment]
Never mind, I can see that you started working on the translation fixes in your PR. I can work on the composer bug, so I am still waiting for info on that one @MonilBhavsar

File: src/pages/iou/HoldReasonPage.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            />
+            <FormProvider
+                formID="moneyHoldReasonForm"
+                submitButtonText={translate('iou.holdRequest')}
+                style={[styles.flexGrow1, styles.ph5]}
+                onSubmit={onSubmit}
+                validate={validate}
+                enabledWhenOffline
+            >
+                <Text style={[styles.textHeadline, styles.mb6]}>{translate('iou.explainHold')}</Text>
```
Comment (s77rt) [inline src/pages/iou/HoldReasonPage.tsx:79]: We shouldn't have set `styles.textHeadline` style for consistency with other pages.

File: src/pages/iou/HoldReasonPage.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    const validate = useCallback((values: FormOnyxValues<typeof ONYXKEYS.FORMS.MONEY_REQUEST_HOLD_FORM>) => {
+        const errors: FormInputErrors<typeof ONYXKEYS.FORMS.MONEY_REQUEST_HOLD_FORM> = ValidationUtils.getFieldRequiredErrors(values, [INPUT_IDS.COMMENT]);
+
+        if (!values.comment) {
+            errors.comment = 'common.error.fieldRequired';
+        }
+
+        return errors;
+    }, []);
```
Comment (alitoshmatov) [inline src/pages/iou/HoldReasonPage.tsx:59]: Quite an edge case but when user A opens hold request reason and another user B settles the request in the same time, the user A presses the hold button but gets no feedback nor reason page closes. ref: https://github.com/Expensify/App/issues/37014

**Comment by sobitneupane** [comment]
https://github.com/Expensify/App/issues/37554 issue is probably a regression from this [PR](https://github.com/Expensify/App/issues/37554).

**Comment by Santhosh-Sellavel** [comment]
This PR causes a regression https://github.com/Expensify/App/issues/37536. Hold Option should not be shown while report while in Scanning state. Also minor alignment issue.

File: src/libs/actions/IOU.ts [hunk 6]
```diff
@@ -3906,6 +3910,102 @@ function getIOUReportID(iou?: OnyxTypes.IOU, route?: MoneyRequestRoute): string
...
+            value: {
+                [createdReportAction.reportActionID]: createdReportAction as ReportAction,
+            },
+        },
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.TRANSACTION}${transactionID}`,
+            value: {
+                comment: {
+                    hold: null,
```
Comment (QichenZhu) [inline src/libs/actions/IOU.ts:3982]: Coming from issue https://github.com/Expensify/App/issues/62602, we should roll back this optimistic data on failure. Fixed in PR https://github.com/Expensify/App/pull/66906.

File: src/libs/actions/IOU.ts [hunk 7]
```diff
@@ -3906,6 +3910,102 @@ function getIOUReportID(iou?: OnyxTypes.IOU, route?: MoneyRequestRoute): string
...
+ */
+function putOnHold(transactionID: string, comment: string, reportID: string) {
+    const createdReportAction = ReportUtils.buildOptimisticHoldReportAction(comment);
+
+    const optimisticData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportID}`,
+            value: {
+                [createdReportAction.reportActionID]: createdReportAction as ReportAction,
```
Comment (FitseTLT) [inline src/libs/actions/IOU.ts:3924]: We should have reset the `pendingAction` of the optimistic `createdReportAction` on successData. Same applies to unHold too.

**Pull request merged**